source_code/main_mcu/src/TIMER/driver_timer.c

```diff
@@ -221,6 +221,54 @@ void timer_set_calendar(uint16_t year, uint16_t month, uint16_t day, uint16_t ho
 #endif
 }
 
+/*! \fn	time(void)
+*   \brief  Get the current unix time
+*
+*   Compute the number of seconds since EPOCH (Jan 1st. 1970), also
+*   commonly known as UNIX time. UNIX time does NOT include LEAP seconds.
+*   See https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xbd_chap04.html#tag_21_04_16
+*   \param  None
+*   \return Seconds since Jan 1st. 1970
+*/
+uint64_t time(void)
+{
+    uint64_t unix_time = 0;
+    int32_t i;
+    uint32_t num_days = 0;
+    calendar_t tm;
+    uint8_t is_curr_year_leap;
+    uint16_t year;
+
+    /* Assuming timer_get_calendar() returns UTC time */
+    timer_get_calendar(&tm);
```

Comment:
that one will give you UTC time... so if you want to convert to UNIX time, should you include the leap seconds? 
current difference is 37 seconds

Reply:
You mean the current difference between UNIX time and UTC time is 27 seconds right? There have been only 27 leap seconds added (However, difference between UTC and TAI is 37 seconds)

So when I am currently testing I am using whatever time you set on the MiniBLE (through Daemon running on host I assume?) which is my local time. I take this local time and add X hours to get UTC time. That is all I do. I don't change seconds. When I compare the TOTP value and rollover with what I get if I show the TOTP on my desktop app (KeePassXC) AND on Google Authenticator on my phone they match exactly. The TOTP value matches up and they roll over at the same time. 

But I _do_ agree with you. We should be subtracting 27 seconds from the UTC time to get the UNIX time. So not sure what is going on. If I do subtract 27 seconds the rollover is of course 27 seconds off based on what I explained above. Login to a real web site still works because they have built in some wiggle room.

What do you think?

Reply:
27 seconds indeed right.
Hmm... your findings are worrying to be honest... i'm not really sure what to think.

Reply:
As discussed offline lets leave as is.